### PR TITLE
Define CH4 RMA synchoronization hooks 

### PR DESCRIPTION
**Pull request by minsii**
In current code, a netmod defines its own RMA synchronization (e.g., flush, fence) in netmod RMA sync functions. If the netmod does not have special HW-RMA, it simply falls back to the AM-based RMA sync routines. In the device layer, an RMA sync API always calls the netmod function.

The above design should be changed in order to add SHM support (i.e., both the shmmod and netmod may define its synchronization). This PR defines four RMA sync hooks for netmod (same hooks will be defined for SHM in another PR). The netmods (ofi and ucx) implement any netmod-specific synchronization (for local or remote completion) in the hooks. The hooks are always triggered in the corresponding fallback AM routine.

Hook API:
1. window remote complete
2. window local complete
3. target remote complete
4. target local complete

Netmod implementation:
- **OFI** use window remote flush for all sync hooks (i.e., waiting on a per-window counter or a global counter)
- **UCX** does not support local flush. It flushes a global worker for window {remote|local} complete, and flushes the target's ep for per-target {remote|local} complete.

We note that, this PR follows the original code design that always issues both HW-sync and RMA-sync. We can optimize the case where no hw-op or no AM-op is issued. This will be addressed in further PR.

See details of the new design in (page 7 is the part of RMA synchronization):
[2018-04-13-ch4-shm.pptx](https://github.com/pmodels/mpich/files/2185856/2018-04-13-ch4-shm.pptx)

This PR is separated from the large SHM/RMA PR #3027 

This PR depends on #3212

### CODE REVIEW

**Comment by minsii** [comment]
test:jenkins/ch4/ofi

**Comment by minsii** [comment]
test:jenkins/ch4/most

**Comment by minsii** [comment]
test:jenkins/ch4/most

**Comment by minsii** [comment]
@pavanbalaji : could you please review it ? This PR defines netmod hooks for RMA sync.

**Comment by hajimefu** [comment]
@minsii For synchronization functions, now that netmod functions (e.g. `MPIDI_NM_mpi_win_fence`) are simply calling CH4R functions, does it make sense to keep these netmod-level functions? I would propose removing them.

In my proposal, `MPIDI_CH4R_mpi_win_fence` will become `MPID_Win_fence` and netmods will simply provide hooks. This structure will solve the Win_fence deadlock issue with per-VNI critical section cleanly. (I think this is what @raffenet suggested as a possible direction.)

My assumption is that netmods are not interested in implementing/optimizing process synchronization part (i.e. barrier in `MPI_Win_fence`, send/recv in `MPI_Win_lock/unlock`.)

**Comment by minsii** [comment]
@hajimefu The reason we keep the synchronization functions is because some netmods might prefer to entirely rewrite the RMA synchronization in `ch4-direct=netmod` mode. 

Can you introduce more about the deadlock issue in per-VNI critical section ?

**Comment by hajimefu** [comment]
@minsii The problem comes from two facts.
1. In per-VNI CS, we have a principle that a per-VNI lock must be taken when calling any netmod function
2. MPI_Win_fence includes barrier

So `MPI_Win_fence` implementation would look like this. (Assuming single VNI for simplicity)
```C
MPID_Win_fence(MPIR_Win *win, ...)
{
    lock(VNI);
    MPIDI_NM_mpi_win_fence(win);
    unlock(VNI);
}

MPIDI_NM_mpi_fence(MPIR_Win *win, ...)
{
    flush_RMA_in_fabric();
    MPIR_Barrier(win->comm);
}
```

This leads to deadlock in `test/mpi/threads/rma/multifence` test where each rank has multiple threads, each of which is calling `MPI_Win_fence` concurrently.

I [presented this issue before](https://anl.app.box.com/s/ziz83mar1glxikh5wtjgmrucbfs6zus6) at the CH4 call. Our proposal then was to split `MPI_Win_fence` into two operations, a) fabric-specific RMA flush part (in OFI, wait for the counter), and b) barrier part. Your hook proposal here is basically equivalent to this "fabric-specific RMA flush part". But in your proposal, netmod synchronization function still includes flush+barrier, so it doesn't solve the deadlock issue.

**Comment by minsii** [comment]
@hajimefu : Sorry for my delayed response. I just switched back from another task. I discussed with @halimamer about this problem. It seems that setting the VNI lock at CH4 fallback functions make more sense than doing it at `MPID_` functions. 

Here are some rules we considered:
1. `MPID_put|win_fence|send|recv` function only decides whether this operation goes through netmod or shm.
2. We should allow a netmod or shm to have full control (including VNI lock) if it chooses to do so.
3. In CH4 fallbacks, we define a set of full implementation including the VNI lock.
4. In current ofi and ucx, the netmod does not need to fully rewrite the operation, but only reuse the CH4 fallbacks with additional netmod-specific sync through hooks.
5. Future netmod or shmmod might want to take 2.

What do you think ?

**Comment by shawnccx** [comment]
I just went through the code change and comments above, `win_fence` deadlock case can be solved if we only set VNI lock on `netmod sync hooks`. This sounds reasonable as this is only netmod-specific part in fence. 

As long as `MPI_Barrier()` is moved out of per-VNI CS, race condition in multi-ranks on multi-threads case can be avoided.

**Comment by minsii** [comment]
To be clear, we proposed adding the per-VNI CS at CH4 fallback routines. For example:
```
MPIDI_CH4R_mpi_win_fence(...) {
    lock(VNI);
    MPIDI_NM_rma_win_cmpl_hook(win);
    unlock(VNI);

    /* AM completion */
    do {
        MPIDI_CH4R_PROGRESS();
    } while (MPIR_cc_get(MPIDI_CH4U_WIN(win, local_cmpl_cnts)) != 0);

    MPIR_Barrier(...);
}
```
`MPIDI_CH4R_mpi_win_fence` is called by `MPIDI_NM_mpi_win_fence` or `MPIDI_SHM_mpi_win_fence` according on the target rank locality. Does this make sense to you ?

**Comment by minsii** [comment]
I assume the `lock(VNI)` should be always called in `MPIDIG_mpi_send`. For netmod send (I see both OFI and UCX customized send), we might want to define a hook (or fallback) to allow the netmod functions to access the CH4 VNI lock. For example:
```
MPIDI_NM_mpi_send(...) {
    if (!MPIDI_OFI_ENABLE_TAGGED) {
         MPIDIG_mpi_send(...);

    MPIDIG_lock_vni(VNI);
    MPIDI_OFI_send(...);
    MPIDIG_unlock_vni(VNI);
}
```
Does it make sense ?

**Comment by hajimefu** [comment]
@minsii So what you are proposing is basically to move lock/unlock responsibility from CH4 to netmods.

We (@shawnccx, @sagarth, and me) had a discussion and here are our opinions.
* Your [proposal for win_fence above](https://github.com/pmodels/mpich/pull/3219#issuecomment-406654398) look good for us.
* But we don't like [the proposal for send](
https://github.com/pmodels/mpich/pull/3219#issuecomment-406679583) for several reasons:
    * It will complicate the netmod design (now a netmod writer has to worry about fine-grain locks for every single netmod function)
    * If this is combined with trylock-enqueue or handoff model, it will generate tons of code duplication. In the current design, since lock and enqueue decision is done at the CH4 (MPID) level, what we enqueue is a netmod-agnostic task structure. But with this design, if the lock acquisition fails in a netmod (with trylock-enqueue model), each netmod has to enqueue its own task structure. I imagine this structure would look quite similar between OFI and UCX (and other) netmods, but they are still different structures. This will end up with code duplication for each netmod function.

So overall, our proposal is to
* Keep the current design (CH4 takes the lock) for small primitives e.g. send/recv/put/get
* Decompose a large netmod function (e.g. win_fence, window creation, etc.) to `MPIDIG` + NM-specific hook and let `MPIDIG` function handle the lock as you proposed. If a netmod really wants to implement its own `win_fence` (=without using `MPIDIG_win_fence`), then that netmod has to take care of the VNI lock inside `NM_mpi_win_fence`.

BTW, is there any reason for changing the design for send/recv/put/get as well? For consistency?

**Comment by minsii** [comment]
It looks like such consistency is hard to maintain for both two-sided and one-sided.

If we give up the consistency part, the RMA design looks  good to me. Just the following sentence is a little confusing to me.
"
Decompose a large netmod function (e.g. win_fence, window creation, etc.) to MPIDIG + NM-specific hook and let MPIDIG function handle the lock as you proposed.
"
I want to clarify that, the `NM-specific hook` should only be called inside MPIDIG (MPIDI_CH4R in current code). Please see the above example of `MPIDI_CH4R_mpi_win_fence`.

**Comment by hajimefu** [comment]
> I want to clarify that, the NM-specific hook should only be called inside MPIDIG (MPIDI_CH4R in current code). Please see the above example of MPIDI_CH4R_mpi_win_fence.

This is what I was thinking too.

Just want to clarify from my side too, that I think this MPIDIG+NMhook design should be applied only to "relatively big" functions, like window creation/destruction and window synchronization. RMA ops (put/get/acc/...) should not adopt this design. So I think it's more of complexity/size of each operation rather than one-sided/two-sided problem.

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets,psm2/all/all/all/debug,default

File: src/mpid/ch4/netmod/ofi/ofi_win.h [hunk 1]
```diff
@@ -1207,4 +1207,24 @@ static inline int MPIDI_NM_mpi_win_free_hook(MPIR_Win * win)
     goto fn_exit;
 }
 
+static inline int MPIDI_NM_rma_win_cmpl_hook(MPIR_Win * win)
```
Comment (hajimefu) [inline src/mpid/ch4/netmod/ofi/ofi_win.h]: `MPL_STATIC_INLINE_PREFIX`?

**Comment by minsii** [comment]
test:jenkins/ch4/most

File: src/mpid/ch4/src/ch4r_win.h [hunk 2]
```diff
@@ -15,6 +15,7 @@
 #include "ch4i_util.h"
 #include <opa_primitives.h>
 #include "mpir_info.h"
+#include "ch4r_symheap.h"
```
Comment (shawnccx) [inline src/mpid/ch4/src/ch4r_win.h]: This line change should not be placed int this commit, right? This commit is for separating function calls under different ch4-direct mode.

Comment (minsii) [reply src/mpid/ch4/src/ch4r_win.h]: I rebased this PR on top of #3212 for compiling & testing. This change should be a part of #3212. Please see my explanation in that PR.

**Comment by shawnccx** [comment]
looks good to me, only need to see tests pass.

**Comment by minsii** [comment]
test:jenkins/ch4/most

**Comment by minsii** [comment]
@shawnccx @hajimefu : I replaced the static inline with `MPL_STATIC_INLINE_PREFIX`, and fixed a bug when `-inline=no` is set.

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/all/all/all/debug,default

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/all/all/all/debug,default

**Pull request merged**